### ISSUE TITLE
Question about `otherProps` in 3.5.0

### ISSUE THREAD

**Issue by prashaantt** [mobxjs/mobx-react#80]
From the README:

```typescript
@inject((allStores) => ({
    userStore: allStores.userStore as IUserStore
}))
class MyComponent extends React.Component<{ userStore?: IUserStore; otherProp: number }, {}> {
    /* etc */
}
```

I was unable to get any props other than those directly injected (`userStore`) to pass through (e.g. `otherProp`). Adding `otherProp` inside `inject` fixes this, but must I now rewrite all other props as such? Is this by design or a bug?

**Comment by mweststrate** [comment]
p props are passed in a second arg. use destructuring or object.assign to
easily pass them on

Op ma 11 jul. 2016 23:23 schreef Prashant Tiwari <notifications@github.com>:

**Comment by prashaantt** [comment]
Got it, thanks. If you'd like, I could send a PR to clarify this in the doc.

**Comment by mweststrate** [comment]
I definitely like that :) Always feel free to improve docs!

Op di 12 jul. 2016 om 00:24 schreef Prashant Tiwari <
notifications@github.com>:

**Comment by ryanatkn** [comment]
After converting some code over to use the new inject decorator I found that I would prefer a slightly different API, one that would behave as @prashaantt expected. I was also confused by the behavior until I read the code. What do you think of making the function passed to `inject` return an object that is merged into the props, rather than return the new props? This is how the Redux `@connect` decorator works. It reduces verbosity by pushing the assign logic up into the HOC's render function, and I think it may be desirable to narrow the stated purpose of the function to only inject props, not map props.

With this proposal, there's a question of which order the props should be resolved. One the one hand, if injected props are copied to the new props first, props passed directly to the component would override them, allowing for example easy mocks in tests and giving more control to parent components. On the other hand if injected props are copied after the new props, the injector function is afforded the same power that the current design has.

Here's what it currently looks like:

```typescript
interface InjectedStores {
  store?: MyStore;
}

interface Props extends InjectedStores {
  otherProp: string;
}

@inject((stores: Stores, nextProps: Props): Props => {
  return assign({store: stores.myStore}, nextProps);
})
@observer
export default MyComponent extends React.Component<Props, {}> {
    ...
```


Here's what the proposal would look like in the common case where the current props are not needed:

```typescript
interface InjectedStores {
  store?: MyStore;
}

interface Props extends InjectedStores {
  otherProp: string;
}

@inject((stores: Stores): InjectedStores => ({store: stores.myStore}))
@observer
export default MyComponent extends React.Component<Props, {}> {
    ...
```

Here's the code right now:

```javascript
function createStoreInjector(grabStoresFn, component) {
  var Injector = React.createClass({
    displayName: "MobXStoreInjector",
    render: function() {
      var newProps = {};
      for (var key in this.props)
        newProps[key] = this.props[key];
      newProps = grabStoresFn(this.context.mobxStores || {}, newProps, this.context);
      return React.createElement(component, newProps);
    }
    ...
```

The proposal would look like this:

```javascript
function createStoreInjector(grabStoresFn, component) {
  var Injector = React.createClass({
    displayName: "MobXStoreInjector",
    render: function() {
      var newProps = {};

      // The injected props could be copied first.
      // This would give parent components more control over the props.
      var injectedStores = grabStoresFn(this.context.mobxStores || {}, this.props, this.context);
      for (var key in injectedStores)
        newProps[key] = injectedStores[key];

      for (var key in this.props)
        newProps[key] = this.props[key];

      // Or the injected props could be copied afterwards.
      // This would give the injector function more control over the props.
      var injectedStores = grabStoresFn(this.context.mobxStores || {}, newProps, this.context);
      for (var key in injectedStores)
        newProps[key] = injectedStores[key];

      return React.createElement(component, newProps);
    }
    ...
```

Thanks!

**Comment by prashaantt** [comment]
@ryanatkn I agree, I think this would be a better design. Having to copy each prop inside the `inject` feels a little non-standard and couples the code very tightly to MobX.

In fact, your first solution was the earlier default behaviour (props > injected), and could be left intact in an improved `inject` that automatically copies over the parent props, while the flexibility to override that behaviour by the current method should also be allowed. That would give the full power back for each component's customisation requirements.

**Comment by prashaantt** [comment]
Although one could argue that the other API for inject: `inject(...varargs)` achieves the same result, and I had begun to favour that for its much less verbosity.

The other thing I would really like to have is some way to make the store names more type-safe, something like `inject(userStore: IUserStore, postStore: IPostStore)`.

**Comment by mweststrate** [comment]
@ryanatkn @prashaantt I agree, this is a bit unexpected. I'll change it in the next version. 

@prashaantt not sure how to make it strong types as you proposes, because we are on the call site and not on the definition side of things?

**Comment by mweststrate** [comment]
@prashaantt the best I could do for strong typing is I think: https://github.com/mobxjs/mobx-react#strongly-typing-inject. It is a bit boilerplaty but the compiler can follow this at least :)

**Comment by prashaantt** [comment]
> the best I could do for strong typing is I think: https://github.com/mobxjs/mobx-react#strongly-typing-inject

I know, that ensures that you're passing the right store type to props. But that's not automatically in sync with the component's actual props as supplied via the interface generic. I agree, there's no way to fix this.

Nevertheless, huge "props" to you for such an awesome library already. Looking forward to the next release. 👍

**Comment by mweststrate** [comment]
Released as `3.5.4`, the returned props from the function are now mixed in (and override) the original `nextProps` props before being passed to the child.
